Bluetooth: Controller: Fix sync reports post terminate and rx disable

cvinayak opened the pull request:
Fix generating periodic advertising reports post sync
terminate under race condition when disabling reporting or
terminating the sync or while performing HCI reset.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix sync reports post terminate and rx disable

Fix generating periodic advertising reports post sync
terminate under race condition when disabling reporting or
terminating the sync or while performing HCI reset.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -6529,6 +6529,7 @@ static void le_per_adv_sync_report(struct pdu_data *pdu_data,
 	uint8_t cte_type = BT_HCI_LE_NO_CTE;
 	struct pdu_adv_com_ext_adv *p;
 	struct pdu_adv_ext_hdr *h;
+	struct ll_sync_set *sync;
```
Given that `sync` is only used once to check a value, consider making this `const`

```suggestion
	const struct ll_sync_set *sync;
```

cvinayak replied:
done

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -6548,6 +6549,18 @@ static void le_per_adv_sync_report(struct pdu_data *pdu_data,
+	 */
+	sync = HDR_LLL2ULL(ftr->param);
+	if (!sync->timeout_reload) {
+		/* NOTE: This is a workaround, to use the field in the context
```
I know this is a workaround.... but it is a bit wired. This is conceptually like using a malloc-ed memory after call to free.
Should it be done this way?

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -6548,6 +6549,18 @@ static void le_per_adv_sync_report(struct pdu_data *pdu_data,
 		return;
 	}
 
+	/* FIXME: make node_rx self sufficient with data when generating so that
```
Is that possible? How to e.g. stop reports if the `sync` is terminated by Host after `node_rx` is enqueued towards ULL/HCI?

ppryga replied:
ok, then it is possible that when this code is executed the `sync` is already released. Are you sure that it will never be overwritten between the release and 'sync->timeout_reload` access, that is done here?

If there is no other way to solve this, there should be an information that `timeout_reload` may not be moved from safe position. You could use `BUILD_ASSERT` and verify if the `timeout_reload` is at fixed offset from beginning of the `lll_sync`

cvinayak committed:
Bluetooth: Controller: Add static build assert to check mem member

Added BUILD_ASSERT definition to check for safe access to
memory pool allocated memory structure member after the
memory has been released back to mem pool.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -6548,6 +6549,22 @@ static void le_per_adv_sync_report(struct pdu_data *pdu_data,
 		return;
 	}
 
+	/* NOTE: The timeout_reload field in the sync context is checked under
+	 *       race condition between HCI Tx and Rx thread wherein a sync
+	 *       terminate was performed which resets the timeout_reload field
+	 *       before releasing the sync context back into its memory pool.
```
Should it be `volatile` then?

cvinayak replied:
It is `volatile`, https://github.com/zephyrproject-rtos/zephyr/blob/40d6fe0ef62403c8a2396f34a38ed191652852a3/subsys/bluetooth/controller/ll_sw/ull_sync_types.h#L21

Thalley replied:
Ah, so if the field is volatile, then the pointer to the struct doesn't need to be? That's good to know :)

cvinayak replied:
The structure's indirect addressing does not change, only the value at the member offset (here `timeout_reload`) is `volatile`.

cvinayak committed:
Bluetooth: Controller: Fix uninitialized param in sync report

Fix uninitialized param field when generating incomplete
no more data to come periodic advertising report.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
Added a commit that fixes an existing bug exposed by changes in this PR.

ppryga-nordic reviewed (commented):
Requested changes are done.

@cvinayak are you sure that it is not possible, that the code will access `sync` that was released and allocated again for a new periodic sync?

cvinayak commented:
> Requested changes are done.
> 
> @cvinayak are you sure that it is not possible, that the code will access `sync` that was released and allocated again for a new periodic sync?

Following is the configuration of co-operative thread:
`CONFIG_BT_HCI_TX_PRIO=7`
`CONFIG_BT_HCI_RX_PRIO=8`

Rx co-operative thread has higher priority of execution than Tx co-operative thread.

After the sync terminate is processed, any reports in race will be processed in Rx thread and HCI report will be generated before next HCI command, i.e. sync create will be processed in the Tx Thread. This should ensure that the report will not be accessing a re-allocated sync context.

ppryga commented:
@cvinayak thank you for explanation.
The workaround is something that in my humble opinion should be temporary.
Proper solution, that does not depend on the unchanged memory of released object, should be provided soon.
I approve the PR only on that condition.

cvinayak commented:
@wopu-ot and @mtpr-ot  Do you have any suggestion on how to handle the race between terminate and reports being generated thereafter. I believe this scenario arises after done event handling was separated out (similar to the complexity that is added in handling connection cancelling for central and connectable advertising disable, solution of using is_stop flags).

mtpr-ot commented:
> @wopu-ot and @mtpr-ot Do you have any suggestion on how to handle the race between terminate and reports being generated thereafter. I believe this scenario arises after done event handling was separated out (similar to the complexity that is added in handling connection cancelling for central and connectable advertising disable, solution of using is_stop flags).

Will take a look and discuss with Wolfgang.

mtpr-ot commented:
@cvinayak, we have seen a similar issue with EBQ testing and HCI reset during CIS tests. HCI RX thread runs the `reset` function, which resets memory pools etc. After the thread is done, the HCI TX thread runs and processes messages already in the `k_fifo`.
This goes bad because the node memory was initialized.
The fix for that was to signal the TX thread using a `k_poll_signal`, which is processed first in the TX thread when signaled, and will empty the `k_fifo`, thereby preventing processing of "dangling" nodes.
Can you explain to what degree what you experience is different from the described scenario?
